crates/sui-core/src/checkpoints/checkpoint_executor/utils.rs

```diff
@@ -0,0 +1,199 @@
+// that get_scheduling_timeout() can be called multiple times in a multithreaded context, but
+// the function is still very cheap to call so this is okay.
+thread_local! {
+    static SCHEDULING_TIMEOUT: once_cell::sync::OnceCell<CheckpointTimeoutConfig> =
+        const { once_cell::sync::OnceCell::new() };
+}
+
+#[cfg(msim)]
+pub fn init_checkpoint_timeout_config(config: CheckpointTimeoutConfig) {
+    SCHEDULING_TIMEOUT.with(|s| {
+        s.set(config).expect("SchedulingTimeoutConfig already set");
+    });
+}
+
+fn get_scheduling_timeout() -> CheckpointTimeoutConfig {
+    fn inner() -> CheckpointTimeoutConfig {
+        let panic_timeout: Option<Duration> = if cfg!(msim) {
+            Some(Duration::from_secs(45))
+        } else {
+            std::env::var("NEW_CHECKPOINT_PANIC_TIMEOUT_MS")
+                .ok()
+                .and_then(|s| s.parse::<u64>().ok())
+                .map(Duration::from_millis)
+        };
+
+        let warning_timeout: Duration = std::env::var("NEW_CHECKPOINT_WARNING_TIMEOUT_MS")
+            .ok()
+            .and_then(|s| s.parse::<u64>().ok())
+            .map(Duration::from_millis)
+            .unwrap_or(Duration::from_secs(5));
+
+        CheckpointTimeoutConfig {
+            panic_timeout,
+            warning_timeout,
+        }
+    }
+
+    SCHEDULING_TIMEOUT.with(|s| *s.get_or_init(inner))
+}
+
+pub(super) fn assert_not_forked(
+    checkpoint: &VerifiedCheckpoint,
+    tx_digest: &TransactionDigest,
+    expected_digest: &TransactionEffectsDigest,
+    actual_effects_digest: &TransactionEffectsDigest,
+    cache_reader: &dyn TransactionCacheRead,
+) {
+    if *expected_digest != *actual_effects_digest {
+        let actual_effects = cache_reader
+            .get_executed_effects(tx_digest)
+            .expect("actual effects should exist");
+
+        // log observed effects (too big for panic message) and then panic.
+        error!(
+            ?checkpoint,
+            ?tx_digest,
+            ?expected_digest,
+            ?actual_effects,
+            "fork detected!"
+        );
+        panic!(
+            "When executing checkpoint {}, transaction {} \
+            is expected to have effects digest {}, but got {}!",
+            checkpoint.sequence_number(),
+            tx_digest,
+            expected_digest,
+            actual_effects_digest,
+        );
+    }
+}
+
+#[derive(Default)]
+pub(super) struct TPSEstimator {
+    last_update: Option<Instant>,
+    transaction_count: u64,
+    tps: f64,
+}
+
+impl TPSEstimator {
+    pub fn update(&mut self, now: Instant, transaction_count: u64) -> f64 {
+        if let Some(last_update) = self.last_update {
+            if now > last_update {
+                let delta_t = now.duration_since(last_update);
+                let delta_c = transaction_count - self.transaction_count;
+                let tps = delta_c as f64 / delta_t.as_secs_f64();
+                self.tps = self.tps * 0.9 + tps * 0.1;
```

Comment:
what is this 0.9 and 0.1 stuff? if it's a moving average can this use `simple_moving_average`?

Reply:
It's an exponential moving average https://en.wikipedia.org/wiki/Exponential_smoothing - I could do an SMA if this proves to be inaccurate. I don't even know what this metric is used for, just trying not to break any dependencies. I checked in PTN that this gives reasonable estimates.